### ISSUE TITLE
[Enhancement] Make Junction's center stick to grid (instead of corners)

### ISSUE THREAD

**Issue by jbsarrodie** [archimatetool/archi#432]
Hi,

When using junctions in conjunction with "Snap to Grid" and OrthogonalAnchors, most connections to/from junctions look ugly if they have bendpoints.

After some investigations, this is caused by:
- The fact that junctions don't use OrthogonalAnchors. This can't be changed because this would look much more ugly (we tested it some years ago)
- And the way junctions are snapped to grid. Like all objects, it is junctions' corners that are snapped to grid, but doing so means that junction's center is never on a grid intersection which in turn means that connections from bendpoints (which are on a grid intersection) can't be exactly horizontal or vertical.

A fix would be to change the way junctions snap to grid (center instead of corners), but I don't know if this is possible...

**Comment by Phillipus** [comment]
> but I don't know if this is possible...

Me neither. ;-)

Needs another deep dive into GEF/Draw2D...

**Comment by jbsarrodie** [comment]
BTW - This is by no way urgent. I just logged it to add it on the (desesperately growing) backlog ;-)

**Comment by rojoweko** [comment]
I experimented a little bit and maybe I found a solution approach for this issue:
if JunctionFigure implements HandleBounds and the method getHandleBounds as follows, a junction will be aligned to the grid with its center:

	public Rectangle getHandleBounds() {
		return new PrecisionRectangle(getBounds().getCenter(), new Dimension());
	}

I cannot tell if this will have any negative side effects, yet, but maybe it will come in handy. ;-)

**Comment by jbsarrodie** [comment]
Thank you, I'll experiment.

From your test, do you confirm that such behavior solves the UI issues we have with junctions ?

**Comment by Phillipus** [comment]
Thanks for taking a look.

This reduces the size of, and repositions. the drag handle which doesn't really solve the underlying issue. Perhaps the real cause is in the SnapToGrid classes?

**Comment by rojoweko** [comment]
@jbsarrodie: Yes, it solves the issue with the bending points, that are aligned to the grid.

**Comment by rojoweko** [comment]
@Phillipus: I agree that this is most likely a framework issue and my approach is merely some kind of workaround that "abuses" the HandleBounds interface. But maybe it is eligible as an acceptable intermediate solution until a proper fix is available. The information you provided looks interesting! I will take a look. Thanks for sharing! ;-)

**Comment by rojoweko** [comment]
I checked that out. Looking good with your approach to derive a new class from SnapToGrid and override the snapRectangle method! I can provide an experimental patch if you want.

**Comment by Phillipus** [comment]
You could paste a snippet for the general logic of it. I think it might also need to apply to SnapToGeometry. Not sure.

**Comment by rojoweko** [comment]
In the overridden snapRectangle method within the new derived ArchiSnapToGrid class I split up the handling respectively for "WEST"/"HORIZONTAL" and "NORTH"/"VERTICAL". The experimental handling for "HORIZONTAL" is as follows:

		if ((snapLocations & HORIZONTAL) != 0 && gridX > 0) {
			double horizontalCorrection = Math.IEEEremainder(rect.getCenter().preciseX()
					- origin.x - 0.5, gridX);
			correction.setPreciseX(correction.preciseX() - horizontalCorrection);
			snapLocations &= ~HORIZONTAL;
		}
"VERTICAL" can be treated likewise with respect to the Y-coordinate.

In the SnapEditPartAdapter class I instanciated the new ArchiSnapToGrid class instead of the SnapToGrid class in the getSnapToHelper() method:

        // Snap to Grid
        val = (Boolean)fEditPart.getViewer().getProperty(ArchiSnapToGrid.PROPERTY_GRID_ENABLED);
        if(val != null && val.booleanValue()) {
            snapStrategies.add(new ArchiSnapToGrid(fEditPart));
        }


I am not sure, if the handling of HORIZONTAL/VERTICAL is 100% correct, but maybe it works as a proof of concept.
Another consideration in this context: elements with odd dimension values have their centers on an integer coordinate value set. This might look better than elements with even dimension values.
